src/parser/vast_parser.js

```diff
@@ -125,254 +142,279 @@ export class VASTParser extends EventEmitter {
-    const lastAddedAd = vastResponse.ads[adsCount - 1];
-    // if in child nodes we have only one ads
-    // and wrapperSequence is defined
-    // and this ads doesn't already have sequence
-    if (
-      adsCount === 1 &&
-      options.wrapperSequence !== undefined &&
-      options.wrapperSequence !== null &&
-      lastAddedAd &&
-      !lastAddedAd.sequence
-    ) {
-      lastAddedAd.sequence = options.wrapperSequence;
-    }
+      const adsCount = ads.length;
+      const lastAddedAd = ads[adsCount - 1];
+      // if in child nodes we have only one ads
+      // and wrapperSequence is defined
+      // and this ads doesn't already have sequence
+      if (
+        adsCount === 1 &&
+        wrapperSequence !== undefined &&
+        wrapperSequence !== null &&
+        lastAddedAd &&
+        !lastAddedAd.sequence
+      ) {
+        lastAddedAd.sequence = wrapperSequence;
+      }
 
-    // vastResponse.ads is an array of Ads which can either be Inline
-    // or wrapper: we need to recursively resolve all the wrappers
-    // The recursion chain is:
-    // parse -> resolveWrappers -> getAndParse -> parse -> resolveWrappers -> ...
-    this.resolveWrappers(vastResponse, options, cb);
+      const resolveWrappersPromises = [];
+
+      ads.forEach(ad => {
+        const resolveWrappersPromise = this.resolveWrappers(
+          ad,
+          wrapperDepth,
+          originalUrl
+        );
+
+        resolveWrappersPromises.push(resolveWrappersPromise);
+      });
+
+      Promise.all(resolveWrappersPromises)
+        .then(unwrappedAds => {
+          const res = this.util.flatten(unwrappedAds);
+          resolve(res);
+        })
+        .catch(err => {
+          console.log(err);
+          reject(err);
+        });
+    });
   }
 
   /**
-   * Resolves the wrappers contained in the given VASTResponse in a recursive way.
-   * Executes the callback with either an error or the fully resolved VASTResponse.
-   * @param    {VASTResponse} vastResponse - An already parsed VASTResponse that may contain some unresolved wrappers.
-   * @param    {Object} options - An optional Object of parameters to be used in the resolving process.
-   * @callback cb
+   * Resolves the wrappers for the given ad in a recursive way.
+   * Returns a Promise which resolves with the unwrapped ad or rejects with an error.
+   * @param  {Ad} ad - An ad to be unwrapped.
+   * @param  {Number} wrapperDepth - The reached depth in the wrapper resolving chain.
+   * @param  {String} originalUrl - The original vast url.
+   * @return {Promise}
    */
-  resolveWrappers(vastResponse, options, cb) {
-    const wrapperDepth = options.wrapperDepth++;
-
-    // Resolve all the wrappers
-    for (let i = 0; i < vastResponse.ads.length; i++) {
-      const ad = vastResponse.ads[i];
+  resolveWrappers(ad, wrapperDepth, originalUrl) {
+    return new Promise((resolve, reject) => {
+      // Going one level deeper in the wrapper chain
+      wrapperDepth++;
```

Comment:
I might have missed something, but `wrapperDepth` is a `Number` right?  
Which means it will always be a `cloned` of `0` if we call `parse(vastXML, { wrapperDepth: 0 })` (e.g: it will aways be equal to `1` and `maxDepth` will never be reached).  
In the previous implementation, the whole object `options` was passed to the function, which means we have a reference of it and `options.wrapperDepth` is properly incremented each time `resolveWrappers` is called?

Reply:
Not really IMHO. 
- `resolveWrappers` receives `wrapperDepth` as a number and increments it. 
- Then, if if needed, it fetches another Ad and calls `parse` passing the Ad and the incremented `wrapperDepth` value. https://github.com/dailymotion/vast-client-js/pull/221/files#diff-207d2ce8628dea4d7e8aca97ef845cb8R337
-  If `parse` will need to call `resolveWrappers` again, it will call it with the incremented `wrapperDepth` value 

Mainly I am not sure I understood what you mean here: "Which means it will always be a cloned of 0 if we call parse(vastXML, { wrapperDepth: 0 })" :)

Reply:
Okay, just double-checked and I misread where the recursivity is done 😄 .  
What I meant:
```javascript
var o = { inc: 0 }
function withObject(obj) {
  obj.inc++
}
withObject(o)
o // Outputs "{ inc: 1 }" because "obj" is a reference of "o" and its value is modified

var i = 0
function withNumber(num) {
  num++
}
withNumber(i)
i // Outputs "0" because num is a copy (clone) of "i" and the initial variable is not modified
```